lib/charms/traefik_k8s/v0/ingress_per_unit.py

```diff
@@ -186,13 +431,102 @@ def is_failed(self, relation: Relation = None):
                 raise RelationDataMismatchError(relation, unit)
         return False
 
+    def get_request(self, relation: Relation):
+        """Get the IngressRequest for the given Relation."""
+        return IngressRequest(self, relation, self._fetch_ingress_data(relation))
+
+    def _fetch_ingress_data(self, relation: Relation, validate=False):
+        """Fetch and validate the databags.
+
+        For the provider side: the application databag.
+        For the requirer side: the unit databag.
+        """
+        this_unit = self.unit
+        this_app = self.app
+
+        if not relation.app or not relation.app.name:
+            # Handle edge case where remote app name can be missing.
+            return {relation.app: {}, this_app: {}, this_unit: {}}
+
+        ingress_data: dict = {}
+        # we start by looking at the provider's app databag
+        if this_unit.is_leader():
+            # only leaders can read their app's data
+            data = relation.data[this_app].get("data")
+            deserialized = {}
+            if data:
+                deserialized = _deserialize_data(data)
+                if validate:
+                    _validate_data(deserialized, INGRESS_PROVIDES_APP_SCHEMA)
+            ingress_data[this_app] = deserialized
+        else:
+            # non-leader units cannot read/write the app databag
+            ingress_data[this_app] = {}
+
+        # then look at the requirer's (thus remote) unit databags
+        remote_units = [
+            obj for obj in relation.data if isinstance(obj, Unit) and obj is not this_unit
+        ]
+
+        for remote_unit in remote_units:
+            remote_data = relation.data[remote_unit].get("data")
+            remote_deserialized = {}
+            if remote_data:
+                remote_deserialized = _deserialize_data(remote_data)
+                if validate:
+                    _validate_data(remote_deserialized, INGRESS_REQUIRES_UNIT_SCHEMA)
```

Comment:
Would it make more sense to have the validation return a bool instead of raising?